File: code_learning/牛客/contest/2024-7-17/D.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
const int N = 4e5 + 7;
const int MOD = 998244353;

ll func(ll x, ll k)
{
    ll T = 1ll << (k + 1);
    ll val = T / 2;
    x++;
    ll res = x / T * val;
    ll r = x % T;
    r -= val;
    if (r > 0)
        res += r;
    return res;
}

void solve()
{
    ll l, r;
    cin >> l >> r;
    ll ans = 0;
    const int LOG = ceil(log2(r));
    for (int i = LOG; i >= 0; i--)
    {
        ll p = (func(r, i) % MOD - func(l - 1, i) % MOD + MOD) % MOD;
        ans = (ans + p) % MOD;
    }
    cout << ans << '\n';
}

int main()
{
    ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);
    int _;
    cin >> _;
    while (_--)
        solve();
    return 0;
}

File: code_learning/牛客/contest/2024-7-17/D.py
"""
利用循环节寻找规律求解该题，但是Python还是太慢了，依然过不了
"""
import sys
from math import ceil, log2
input = lambda: sys.stdin.readline().rstrip()
mod = 998244353

def func(x, k):
    T = 1 << (k + 1)
    val = T // 2
    x += 1
    res = x // T * val
    r = x % T
    r -= val
    if r > 0:
        res += r
    return res
dp = {}
for _ in range(int(input())):
    ans = 0
    l, r = map(int, input().split())
    log = ceil(log2(r))
    for i in range(log + 1):
        if (r, i) not in dp.keys():
            res1 = func(r, i)
            dp[(r, i)] = res1
        else:
            res1 = dp[(r, i)]
        if (l - 1, i) not in dp.keys():
            res2 = func(l - 1, i)
            dp[(l - 1, i)] = res2
        else:
            res2 = dp[(l - 1, i)]
        p = (res1 % mod - res2 % mod + mod) % mod
        ans = (ans + p) % mod
    sys.stdout.write(str(ans) + '\n')